Bug in own typings in 5.2.0

<!-- Please don't delete this template otherwise your issue will be closed immediately -->
<!-- Before creating an issue please make sure you are using the latest version of webpack. -->

### Bug report

Looks like we have a problem in typings.

<!-- Please ask questions on discussions, StackOverflow or the webpack Gitter. -->
<!-- https://github.com/webpack/webpack/discussions -->
<!-- https://stackoverflow.com/questions/ask?tags=webpack -->
<!-- https://gitter.im/webpack/webpack -->
<!-- Issues which contain questions or support requests will be closed. -->

### Actual Behavior

Errors in tsc output

```bash
node_modules/terser-webpack-plugin/types/index.d.ts(9,38): error TS2307: Cannot find module 'uglify-js' or its corresponding type declarations.
node_modules/terser-webpack-plugin/types/index.d.ts(10,33): error TS2307: Cannot find module '@swc/core' or its corresponding type declarations.
node_modules/terser-webpack-plugin/types/utils.d.ts(4,44): error TS2307: Cannot find module 'uglify-js' or its corresponding type declarations.
node_modules/terser-webpack-plugin/types/utils.d.ts(5,38): error TS2307: Cannot find module 'uglify-js' or its corresponding type declarations.
node_modules/terser-webpack-plugin/types/utils.d.ts(6,33): error TS2307: Cannot find module '@swc/core' or its corresponding type declarations.
```

<!-- Explain exactly how it behave -->

### Expected Behavior

No type errors
<!-- "It should work" is not a helpful explanation -->
<!-- Explain exactly how it should behave -->

### How Do We Reproduce?

Just try to run `node_modules/.bin/tsc` in a project with terser-webpack-plugin 5.2.0

Looks like we need to install all possible minifiers. But it looks quite strange. Why do I need to install swc, if I use esbuild?)

My tsconfig is:

```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "tslib": [
        "./node_modules/tslib/tslib.d.ts"
      ],
      "*": [
        "src/*"
      ]
    },
    "esModuleInterop": true,
    "declaration": false,
    "declarationMap": false,
    "moduleResolution": "node",
    "module": "commonjs",
    "incremental": true,
    "inlineSourceMap": true,
    "inlineSources": true,
    "alwaysStrict": true,
    "pretty": false,
    "jsx": "react-jsxdev",
    "strict": true,
    "strictNullChecks": true,
    "target": "es6",
    "noEmitHelpers": true,
    "importHelpers": true,
    "noUncheckedIndexedAccess": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "strictPropertyInitialization": false,
    "lib": [
      "es6",
      "es2016",
      "es2017",
      "dom"
    ]
  },
  "exclude": [
    "node_modules"
  ],
  "compileOnSave": false
}
```

The optimization section from my webpack config:

```js
optimization: {
    concatenateModules: true,
    minimizer: [
      new TerserPlugin({
        minify: TerserPlugin.esbuildMinify,
      }),
    ],
  },
```

As a solution, you can add esbuild, @swc/core and uglify-js typings to dependencies.

<!-- A great way to do this is to provide your configuration via a GitHub repository -->
<!-- The most helpful is a minimal reproduction with instructions on how to reproduce -->
<!-- Repositories with too many files or large `webpack.config.js` files are not suitable -->
<!-- Please only add small code snippets directly into this issue -->
<!-- https://gist.github.com is a good place for longer code snippets -->
<!-- If your issue is caused by a plugin or loader, please create an issue on the loader/plugin repository instead -->

### Please paste the results of `npx webpack-cli info` here, and mention other relevant information

```
System:
    OS: macOS 10.15.7
    CPU: (8) x64 Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
    Memory: 619.85 MB / 16.00 GB
  Binaries:
    Node: 14.17.3 - ~/n/bin/node
    Yarn: 1.22.0 - ~/.yarn/bin/yarn
    npm: 6.14.13 - ~/n/bin/npm
  Browsers:
    Chrome: 92.0.4515.159
    Firefox: 88.0.1
    Safari: 14.0.1
    Safari Technology Preview: 14.1
  Packages:
    compression-webpack-plugin: 8.0.1 => 8.0.1
    terser-webpack-plugin: 5.2.0 => 5.2.0
    webpack: 5.51.1 => 5.51.1
    webpack-bundle-analyzer: 4.4.2 => 4.4.2
    webpack-cli: 4.8.0 => 4.8.0
    webpack-merge: 5.8.0 => 5.8.0
    webpack-npm-dependencies-analyzer: 1.0.1 => 1.0.1
    webpack-stats-plugin: 1.0.3 => 1.0.3
```

right, this version breaks , i got one line more then above

```
node_modules/terser-webpack-plugin/types/index.d.ts:11:37 - error TS2307: Cannot find module 'esbuild' or its corresponding type declarations.

export type EsbuildOptions = import("esbuild").TransformOptions;
```

--- a/src/index.js
+++ b/src/index.js
@@ -22,11 +22,11 @@ import { minify as minimize } from "./minify";
 /** @typedef {import("webpack").Compilation} Compilation */
 /** @typedef {import("webpack").WebpackError} WebpackError */
 /** @typedef {import("webpack").Asset} Asset */
-/** @typedef {import("terser").ECMA} TerserECMA */
-/** @typedef {import("terser").MinifyOptions} TerserOptions */
-/** @typedef {import("uglify-js").MinifyOptions} UglifyJSOptions */
-/** @typedef {import("@swc/core").JsMinifyOptions} SwcOptions */
-/** @typedef {import("esbuild").TransformOptions} EsbuildOptions */
+/** @typedef {import("./utils.js").TerserECMA} TerserECMA */
+/** @typedef {import("./utils.js").TerserOptions} TerserOptions */
+/** @typedef {import("./utils.js").UglifyJSOptions} UglifyJSOptions */
+/** @typedef {import("./utils.js").SwcOptions} SwcOptions */
+/** @typedef {import("./utils.js").EsbuildOptions} EsbuildOptions */
 /** @typedef {Object.<any, any>} CustomOptions */
 /** @typedef {import("jest-worker").Worker} JestWorker */
 /** @typedef {import("source-map").RawSourceMap} RawSourceMap */

--- a/src/utils.js
+++ b/src/utils.js
@@ -1,6 +1,7 @@
 /** @typedef {import("source-map").RawSourceMap} RawSourceMap */
 /** @typedef {import("terser").FormatOptions} TerserFormatOptions */
 /** @typedef {import("terser").MinifyOptions} TerserOptions */
+/** @typedef {import("terser").ECMA} TerserECMA */
 /** @typedef {import("uglify-js").OutputOptions} UglifyJSOutputOptions */
 /** @typedef {import("uglify-js").MinifyOptions} UglifyJSOptions */
 /** @typedef {import("@swc/core").JsMinifyOptions} SwcOptions */

--- a/src/utils.js
+++ b/src/utils.js
@@ -2,6 +2,7 @@
 /** @typedef {import("terser").FormatOptions} TerserFormatOptions */
 /** @typedef {import("terser").MinifyOptions} TerserOptions */
 /** @typedef {import("terser").ECMA} TerserECMA */
+// TODO do not forget to add `// @ts-ignore` on optional peer dependencies to avoid problem with `TS2307: Cannot find module 'module' or its corresponding type declarations.`
 /** @typedef {import("uglify-js").OutputOptions} UglifyJSOutputOptions */
 /** @typedef {import("uglify-js").MinifyOptions} UglifyJSOptions */
 /** @typedef {import("@swc/core").JsMinifyOptions} SwcOptions */

--- a/types/index.d.ts
+++ b/types/index.d.ts
@@ -251,7 +251,7 @@ declare class TerserPlugin<
     | UglifyJSMinimizer
     | SwcMinimizer
     | EsbuildMinimizer
-    | CustomMinimizer
+    | CustomMinimizer = TerserMinimizer
 > {
   /**
    * @private

--- a/types/index.d.ts
+++ b/types/index.d.ts
@@ -4,11 +4,11 @@ export type Compiler = import("webpack").Compiler;
 export type Compilation = import("webpack").Compilation;
 export type WebpackError = import("webpack").WebpackError;
 export type Asset = import("webpack").Asset;
-export type TerserECMA = import("terser").ECMA;
-export type TerserOptions = import("terser").MinifyOptions;
-export type UglifyJSOptions = import("uglify-js").MinifyOptions;
-export type SwcOptions = import("@swc/core").JsMinifyOptions;
-export type EsbuildOptions = import("esbuild").TransformOptions;
+export type TerserECMA = import("./utils.js").TerserECMA;
+export type TerserOptions = import("./utils.js").TerserOptions;
+export type UglifyJSOptions = import("./utils.js").UglifyJSOptions;
+export type SwcOptions = import("./utils.js").SwcOptions;
+export type EsbuildOptions = import("./utils.js").EsbuildOptions;
 export type CustomOptions = any;
 export type JestWorker = import("jest-worker").Worker;
 export type RawSourceMap = import("source-map").RawSourceMap;
@@ -119,11 +119,11 @@ export type PickMinimizerImplementationAndOptions<T> =
 /** @typedef {import("webpack").Compilation} Compilation */
 /** @typedef {import("webpack").WebpackError} WebpackError */
 /** @typedef {import("webpack").Asset} Asset */
-/** @typedef {import("terser").ECMA} TerserECMA */
-/** @typedef {import("terser").MinifyOptions} TerserOptions */
-/** @typedef {import("uglify-js").MinifyOptions} UglifyJSOptions */
-/** @typedef {import("@swc/core").JsMinifyOptions} SwcOptions */
-/** @typedef {import("esbuild").TransformOptions} EsbuildOptions */
+/** @typedef {import("./utils.js").TerserECMA} TerserECMA */
+/** @typedef {import("./utils.js").TerserOptions} TerserOptions */
+/** @typedef {import("./utils.js").UglifyJSOptions} UglifyJSOptions */
+/** @typedef {import("./utils.js").SwcOptions} SwcOptions */
+/** @typedef {import("./utils.js").EsbuildOptions} EsbuildOptions */
 /** @typedef {Object.<any, any>} CustomOptions */
 /** @typedef {import("jest-worker").Worker} JestWorker */
 /** @typedef {import("source-map").RawSourceMap} RawSourceMap */
@@ -251,7 +251,7 @@ declare class TerserPlugin<
     | UglifyJSMinimizer
     | SwcMinimizer
     | EsbuildMinimizer
-    | CustomMinimizer = TerserMinimizer
+    | CustomMinimizer
 > {
   /**
    * @private

--- a/types/utils.d.ts
+++ b/types/utils.d.ts
@@ -1,6 +1,7 @@
 export type RawSourceMap = import("source-map").RawSourceMap;
 export type TerserFormatOptions = import("terser").FormatOptions;
 export type TerserOptions = import("terser").MinifyOptions;
+export type TerserECMA = import("terser").ECMA;
 export type UglifyJSOutputOptions = import("uglify-js").OutputOptions;
 export type UglifyJSOptions = import("uglify-js").MinifyOptions;
 export type SwcOptions = import("@swc/core").JsMinifyOptions;
@@ -42,6 +43,7 @@ export type ExtractedComments = Array<string>;
 /** @typedef {import("source-map").RawSourceMap} RawSourceMap */
 /** @typedef {import("terser").FormatOptions} TerserFormatOptions */
 /** @typedef {import("terser").MinifyOptions} TerserOptions */
+/** @typedef {import("terser").ECMA} TerserECMA */
 /** @typedef {import("uglify-js").OutputOptions} UglifyJSOutputOptions */
 /** @typedef {import("uglify-js").MinifyOptions} UglifyJSOptions */
 /** @typedef {import("@swc/core").JsMinifyOptions} SwcOptions */

--- a/types/utils.d.ts
+++ b/types/utils.d.ts
@@ -2,9 +2,13 @@ export type RawSourceMap = import("source-map").RawSourceMap;
 export type TerserFormatOptions = import("terser").FormatOptions;
 export type TerserOptions = import("terser").MinifyOptions;
 export type TerserECMA = import("terser").ECMA;
+// @ts-ignore
 export type UglifyJSOutputOptions = import("uglify-js").OutputOptions;
+// @ts-ignore
 export type UglifyJSOptions = import("uglify-js").MinifyOptions;
+// @ts-ignore
 export type SwcOptions = import("@swc/core").JsMinifyOptions;
+// @ts-ignore
 export type EsbuildOptions = import("esbuild").TransformOptions;
 export type ExtractCommentsOptions =
   import("./index.js").ExtractCommentsOptions;

